### DISCUSSION TITLE
Feedback portal

### DISCUSSION THREAD

**Discussion by agx** [flatpak/xdg-desktop-portal#1150] [New Portals]
Apps on mobile devices want to submit haptic, led and audio feedback to the user. For that several applications are using [libfeedback/feedbackd](https://source.puri.sm/Librem5/feedbackd). The [DBus API](https://source.puri.sm/Librem5/feedbackd/-/blob/main/data/org.sigxcpu.Feedback.xml) was initially designed to match the portals request/response/cancel pattern so I'd like to add a `org.freedesktop.portal.Feedback` portal that allows for sandboxed apps to submit events to the feedback daemon. Applications using libfeedback would then use the portal transparently instead of connecting to feedbackd directly.

I looked through the MRs and it seems there's no formal process to be followed but wanted to check upfront.

Related downstream issue: https://source.puri.sm/Librem5/feedbackd/-/issues/31

```xml
<?xml version="1.0"?>
<node name="/" xmlns:doc="http://www.freedesktop.org/dbus/1.0/doc.dtd">
  <!--
      org.freedesktop.portal.Feedback:
      @short_description: Portal for emitting audio/led/haptic feedback

      This simple interface lets sandboxed applications emit audio/led/haptic
      feedback.

      This documentation describes version 1 of this interface.
  -->
  <interface name="org.freedesktop.portal.Feedback">
    <!--
        TriggerFeedback:
        @event: The event name from the Event naming spec
        @hints: Additional hints (e.g. to prefer haptic feedback over audio
                 feedback on button press events)
        @timeout: When the feedbacks for this event should end latest in seconds.
                  The special values '-1' (just run each feedback once) and '0' (endless loop) 
                  xare also supported.
        @handle: Object path for the #org.freedesktop.portal.Request object representing this call

        Asks to set trigger feedback for the given event.

        Supported keys in the @hints vardict include:
	...
    -->
    <method name="TriggerFeedback">
      <arg type="s" name="uri" direction="in"/>
      <arg type="a{sv}" name="hints" direction="in"/>
      <arg type="o" name="handle" direction="out"/>
    </method>

    <property name="version" type="u" access="read"/>
  </interface>
</node>
```

If the client wants to end feedback early (e.g. in the case of a picked up phone call) it could just close the request object or we could have an explicit `EndFeedback` call. Would be great to get some input what the preferred way is?
The end of the feedback would be indicated by requests `Response` signal.

**Comment by swick** [comment]
Looking at the [events](https://source.puri.sm/Librem5/feedbackd/-/blob/main/doc/Event-naming-spec-0.0.0.md#standard-event-names) almost all of them seem to be for the system and not apps so why would we want a portal for this?

**Comment by agx** [comment]
@swick some of them are system but most of the current ones are in fact used by applications:

-  `phone-*` (phone and soft-phone (e.g. sip) applications, e.g. gnome-calls)
-  `message-*` (sms, instant messaging and email apps)
- `alarm-clock-elapsed`, `timeout-completed` (alarm clocks, stop watch)
- `camera-*` (camera apps)
- `button-*` (toolkit)

**Comment by jadahl** [comment]
> * `phone-*` (phone and soft-phone (e.g. sip) applications, e.g. gnome-calls)
> * `message-*` (sms, instant messaging and email apps)

Could you expand how they use these signals? Won't a phone call or SMS application get notified about this via the same APIs it makes calls and reads SMS?

And how do instant messaging application, which are likely to handle all the communication with their servers themselves, using these?

> * `alarm-clock-elapsed`, `timeout-completed` (alarm clocks, stop watch)

How do alarm applications use these?

> * `camera-*` (camera apps)

focus/shutter I would assume to be communicated via PipeWire with some libcamera way.

> * `button-*` (toolkit)

What buttons do regular applications need that doesn't come via regular input event ways?

**Comment by agx** [comment]
> > * `phone-*` (phone and soft-phone (e.g. sip) applications, e.g. gnome-calls)
> > * `message-*` (sms, instant messaging and email apps)
> 
> Could you expand how they use these signals? Won't a phone call or SMS application get notified about this via the same APIs it makes calls and reads SMS?

Note that this is not about signalling *to* an application that something happened but rather a way *for* an application to trigger feedback to the user. This frees application from figuring out what hw is available (e.g. is there an haptic motor, if so what is the correct drive strength, what is the color of the status led, is it multicolor,  which led to blink if there are several depending on the device orientation¹, what is the maximum noisiness the user wants atm, ...).

So when the phone application detects an incoming call (e.g. via MM or SIP (the later can work inside a flatpak already)) it emits `TriggerFeedback("phone-incoming-call", ...)` which would cause feedbackd to play the ring tone, emit haptic feedback, etc based on the theme configuration, device capabilities and the users current prefs. When the user picks up, the phone app cancels the request and the ringing ends. How this is implemented is of course desktop / mobile env specific. Feedbackd is just one possible implementation.

> 
> And how do instant messaging application, which are likely to handle all the communication with their servers themselves, using these?
> 
> > * `alarm-clock-elapsed`, `timeout-completed` (alarm clocks, stop watch)
> 
> How do alarm applications use these?

https://sources.debian.org/src/gnome-clocks/43.0-1/src/alarm-item.vala/?hl=113#L113 - apps would simply invoke libfeedback instead of gsound if they want to trigger haptic/led feedback as well.

> 
> > * `camera-*` (camera apps)
> 
> focus/shutter I would assume to be communicated via PipeWire with some libcamera way.

Applications emit this when the user clicks the shutter button, e.g. https://sources.debian.org/src/cheese/43~alpha-1/src/cheese-window.vala/?hl=683#L683  - apps would simply invoke libfeedback instead of gsound if f they want to trigger haptic/led feedback as well.

> 
> > * `button-*` (toolkit)
> 
> What buttons do regular applications need that doesn't come via regular input event ways?

An application might want to have a sound emitted on every button click (similar to what libcanberra allows you to do by ways of the canberra-gtk module.

¹) the device orientation bit isn't implemented yet

**Comment by jadahl** [comment]
Thanks for the explanation, that makes it all clearer. Calling it "events" made me think they were events from the system, not *to* the system.

> bell-terminal: The sound to use as a terminal bell.

This is currently handled by a request on `gtk_surface1`. It was added there because the bell is sometimes audible, and sometimes visual, where e.g. the window it was triggered from flashes. This is part of the accessibility feature set. With that in mind, I'm not sure how well it fits here, we'd need to plumb that request via feedbackd to the display server some how.

Anyhow...

How do you imagine the permission model for feedback to look like? Should applications just request feedback and e.g. the first time for every type of feedback the user would see a query, should it be always-allow with an option to deny via e.g. Settings, or should the app request permission to perform a certain set of feedback?

> If the client wants to end feedback early (e.g. in the case of a picked up phone call) it could just close the request object or we could have an explicit EndFeedback call.

The Request object, being usually used as a handle for the app requesting permission to something, or a privileged resource, I can see why it'd be tempting to reuse, but I'm not sure it's the best idea, as it'd make it harder to extend for things are specific to feedback unless it can only be sent when a request ended. For example, one wouldn't be able to say "why" it closed, or we wouldn't be able to emit signals other than when the request was "done".

An option is to use "Session" for feedback that are long lived.

**Comment by swick** [comment]
All of those have no reason to be ever used by normal apps:

> battery-low: The Event used when the battery is low (below 20%, for example).
> power-unplug-battery-low: The power cable has been unplugged and the battery level is low.
> battery-caution: The event used when the battery is nearing exhaustion (below 40%, for example).
> battery-full:	The event used when the battery is fully loaded up.
> device-added: The event used when a device has become available to the desktop, i.e. due to USB plugging.
> power-plug: The power cable has been plugged in.
> power-unplug: The power cable has been unplugged.
> theme-demo: A event that should be played for demoing this theme. Usually this should just be an alias for a very representative sound (such as a incoming phone call) of a theme that would work nicely as a demo event for a theme in the theme selector dialog.

This sounds like for on-screen keyboards, also not normal apps.:

> button-pressed:	The event used when a button is pressed.
> button-released:  The event used when a button is released.

Only relevant for the compositor, not normal apps:

> window-close:     The sound used when an existing window is closed.


Those sound like they should be the result of a notification:

> message-new-instant: The event used when a new IM is received.
> message-new-sms:  The event used when a new sms is received.
> message-new-email:  The event used when a new email is received.
> message-missed-email: The event used when an email was received but not seen by the user.
> message-missed-instant: The event used when a instant message was received but not seen by the user.
> message-missed-notification: The event used when a notification was shown but not seen by the user.
> message-missed-sms: The event used when a sms message was received but not seen by the user.

Then there are a few remaining which actually look like they might make sense for apps to have access to. All I'm saying here is that this seems really generic and you could improve e.g. the notification portal and backend to make use of this first for example without needing this portal.

**Comment by agx** [comment]
> Thanks for the explanation, that makes it all clearer. Calling it "events" made me think they were events from the system, not _to_ the system.
> 
> > bell-terminal: The sound to use as a terminal bell.
> 
> This is currently handled by a request on `gtk_surface1`. It was added there because the bell is sometimes audible, and sometimes visual, where e.g. the window it was triggered from flashes. This is part of the accessibility feature set. With that in mind, I'm not sure how well it fits here, we'd need to plumb that request via feedbackd to the display server some how.

Happy to skip that one then.

> 
> Anyhow...
> 
> How do you imagine the permission model for feedback to look like? Should applications just request feedback and e.g. the first time for every type of feedback the user would see a query, should it be always-allow with an option to deny via e.g. Settings, or should the app request permission to perform a certain set of feedback?

I'd preferably allow it by default with an option to deny. I'll also separate the possible events into "system" ones (e.g. as used by a compositor, mobile or desktop shell or something like g-s-d) and the ones for applications and only the later would be allowed from the sandbox.

> 
> > If the client wants to end feedback early (e.g. in the case of a picked up phone call) it could just close the request object or we could have an explicit EndFeedback call.
> 
> The Request object, being usually used as a handle for the app requesting permission to something, or a privileged resource, I can see why it'd be tempting to reuse, but I'm not sure it's the best idea, as it'd make it harder to extend for things are specific to feedback unless it can only be sent when a request ended. For example, one wouldn't be able to say "why" it closed, or we wouldn't be able to emit signals other than when the request was "done".
> 
> An option is to use "Session" for feedback that are long lived.

That's good advice, thanks! Another alternative would be to use an id similar to what is done for notifications.

**Comment by jadahl** [comment]
> > button-pressed:	The event used when a button is pressed.
> > button-released:  The event used when a button is released.
> 
> Only relevant for the compositor, not normal apps:

Custom OSK's (e.g. the bank login on my phone) could use these.




> > message-new-instant: The event used when a new IM is received.
> > message-new-sms:  The event used when a new sms is received.
> > message-new-email:  The event used when a new email is received.
> > message-missed-email: The event used when an email was received but not seen by the user.
> > message-missed-instant: The event used when a instant message was received but not seen by the user.
> > message-missed-notification: The event used when a notification was shown but not seen by the user.
> > message-missed-sms: The event used when a sms message was received but not seen by the user.
> 
> Then there are a few remaining which actually look like they might make sense for apps to have access to. All I'm saying here is that this seems really generic and you could improve e.g. the notification portal and backend to make use of this first for example without needing this portal.

These do indeed sound like something that should be part of the notification portal.

The incoming call one I'm not sure about though, it's a lot more synchronous than notifications, which are sent away without even knowing if they ever were shown.

**Comment by agx** [comment]
> All of those have no reason to be ever used by normal apps:
> 
> > battery-low: The Event used when the battery is low (below 20%, for example).
> > power-unplug-battery-low: The power cable has been unplugged and the battery level is low.
> > battery-caution: The event used when the battery is nearing exhaustion (below 40%, for example).
> > battery-full:	The event used when the battery is fully loaded up.
> > device-added: The event used when a device has become available to the desktop, i.e. due to USB plugging.
> > power-plug: The power cable has been plugged in.
> > power-unplug: The power cable has been unplugged.
> > theme-demo: A event that should be played for demoing this theme. Usually this should just be an alias for a very representative sound (such as a incoming phone call) of a theme that would work nicely as a demo event for a theme in the theme selector dialog.

That's true. As noted in another reply I'll make that separation more
obvious.

> 
> This sounds like for on-screen keyboards, also not normal apps.:
> 
> > button-pressed:	The event used when a button is pressed.
> > button-released:  The event used when a button is released.

And e.g.keypads on phone apps dialpad or the "send" button of a IM
application and there's also the "toolkit" case where a toolkit might
want to emit feedback for every button pressed.

> 
> Only relevant for the compositor, not normal apps:
> 
> > window-close:     The sound used when an existing window is closed.
> 
> Those sound like they should be the result of a notification:
> 
> > message-new-instant: The event used when a new IM is received.
> > message-new-sms:  The event used when a new sms is received.
> > message-new-email:  The event used when a new email is received.
> > message-missed-email: The event used when an email was received but not seen by the user.
> > message-missed-instant: The event used when a instant message was received but not seen by the user.
> > message-missed-notification: The event used when a notification was shown but not seen by the user.
> > message-missed-sms: The event used when a sms message was received but not seen by the user.

> Then there are a few remaining which actually look like they might make sense for apps to have access to. All I'm saying here is that this seems really generic and you could improve e.g. the notification portal and backend to make use of this first for example without needing this portal.

You're right that we also need to improve the notification portal for
mobile. This (among other things) includes having the ability to emit
event based feedback¹ but there are cases where there's not necessarily
a notification (in the sense of the xdg [Desktop Notification
Spec](https://specifications.freedesktop.org/notification-spec/notification-spec-latest.html) ) like when an IM app is in the foreground and has input focus but still
wants to indicate "new IM" via a haptic buzz or when you have an
ongoing phone call and there's a 2nd call incoming (or the button
press case from above). I also expect less desktop notifications when
in app notifications are used more due to AdwToast and apps gaining the
ability to indicate outstanding in app messages e.g. in the shell
launchers. 

¹) I'm currently looking at improving these in phosh and once it covers the
interesting cases will look at extending the xdg-spec which I assume is a
prereq for the portal.

**Comment by jadahl** [comment]
> will look at extending the xdg-spec which I assume is a
> prereq for the portal.

I wouldn't say that; the xdg spec is just one potential ways to talk to an underlying layer for a notification portal backend, and applications should not use the xdg spec directly due to issues related to sandboxing. For example xdg-desktop-portal-gtk uses `org.gtk.Notifications`.

**Comment by agx** [comment]
> > will look at extending the xdg-spec which I assume is a
> > prereq for the portal.
> 
> I wouldn't say that; the xdg spec is just one potential ways to talk to an underlying layer for a notification portal backend, and applications should not use the xdg spec directly due to issues related to sandboxing. For example xdg-desktop-portal-gtk uses `org.gtk.Notifications`.

Ah...that's good to know. I'll need to sort out the non sandbox side first but that indeed could make things simpler. Thanks!

**Comment by Mikenux** [comment]
What is left for this portal? Actions (message-sent-instant) and Input Events (button-pressed and button-released)?

> Custom OSK's (e.g. the bank login on my phone) could use these.

I can't imagine having different themes when typing on a keyboard. I think such a thing should be stable and user-defined on the system side.

What is this custom OSK? Because I think the keyboard should be stable and something shown by the system.

> And e.g.keypads on phone apps dialpad or the "send" button of a IM
> application and there's also the "toolkit" case where a toolkit might
> want to emit feedback for every button pressed.

If the keyboard is custom, same as above.

I agree that for certain actions, like the send button, having a custom sound can be nice. However, this can be abused. That said, if there is abuse, the user can remove the app. However, do "we" want portals that allow abuse without a prevention mechanism or portals that avoid such abuse so that users can expect the application to work properly (at least as much as possible)?
